### PR TITLE
feat(oci): add support for referencing an input image by digest

### LINKED ISSUE #3571: feat: Support scanning by digest or platform when providing an OCI layout as --input
I tried scanning an OCI multi-arch image using the Trivi version 0.36.1

I downloaded this image like this:
`skopeo copy --format=oci --dest-tls-verify=false --all docker://alpine:3.11.0 oci:alpine:3.11.0`

After download I have these files in my OCI layout

**alpine/index.json**
```json
{
    "schemaVersion": 2,
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.index.v1+json",
            "digest": "sha256:6905f8b9ecf995010b24b870e3e21b9893a3c2dbaa173d88da78648377c0e8ac",
            "size": 1542,
            "annotations": {
                "org.opencontainers.image.ref.name": "3.11.0"
            }
        }
    ]
}
```

**alpine/blobs/sha256/6905f8b9ecf995010b24b870e3e21b9893a3c2dbaa173d88da78648377c0e8ac**
```json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.index.v1+json",
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:82746c257a7fa34c0da886a0b4a1345512079f5dcaf4312f8401c7365a4681ea",
            "size": 405,
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:0f928aa4adc59f27bd5eab1554b0a026b419d66b744c284d5738759e3f921b33",
            "size": 405,
            "platform": {
                "architecture": "arm",
                "os": "linux",
                "variant": "v6"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:ef22c291783c6af352f3e1b91d97857d74c78eed0a8ca0e0ab8df1f6d630c542",
            "size": 405,
            "platform": {
                "architecture": "arm",
                "os": "linux",
                "variant": "v7"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:2d26cc2816cb6136dfe608dc48594a1ddeb7562c16e1a2aaf34a8906a433d0dd",
            "size": 405,
            "platform": {
                "architecture": "arm64",
                "os": "linux",
                "variant": "v8"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:7f2e55dfee18c656ef6d0c8fd44807e5d916ea95db979e636f18cc84075e5951",
            "size": 405,
            "platform": {
                "architecture": "386",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:c3688e845f9c761a591877c828c2f2b6f31f6e1d8cf4aee60fb5ef6675ec41ee",
            "size": 405,
            "platform": {
                "architecture": "ppc64le",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:fa7f18e151ecaeac2c6d86943a9810fbcf383f9acc19951c16203f57af5b35ab",
            "size": 405,
            "platform": {
                "architecture": "s390x",
                "os": "linux"
            }
        }
    ]
}
```

I tried calling the command:

`trivy image --platform linux/arm64 --input alpine:3.11.0`

This command scans the first manifest inside the index descriptor for my image which is for linux/amd64, not linux/arm64 as I would have expected.

Also, I noticed that specifying an invalid platform does not result in an error. To following command scans successfully the image for linux/amd64 platform again.

`trivy image --platform badOs/badArch --input alpine:3.11.0`

As a workaround I have also tried to scan by a specific manifest digest. I followed the syntax model specified in this issue that was fixed #3075. So, I tried the following command: 
`trivy image --input alpine@sha256:2d26cc2816cb6136dfe608dc48594a1ddeb7562c16e1a2aaf34a8906a433d0dd`

I received the following error:
```
2023-02-07T12:32:34.892+0200    INFO    Vulnerability scanning is enabled
2023-02-07T12:32:34.892+0200    INFO    Secret scanning is enabled
2023-02-07T12:32:34.892+0200    INFO    If your scanning is slow, please try '--security-checks vuln' to disable secret scanning
2023-02-07T12:32:34.892+0200    INFO    Please see also https://aquasecurity.github.io/trivy/v0.36/docs/secret/scanning/#recommendation for faster secret detection
2023-02-07T12:32:34.892+0200    FATAL   image scan error: scan error: unable to initialize a scanner: unable to initialize the archive scanner: 2 errors occurred:
        * unable to open alpine:sha256:82746c257a7fa34c0da886a0b4a1345512079f5dcaf4312f8401c7365a4681ea as a Docker image: unable to open the file: open alpine:sha256:82746c257a7fa34c0da886a0b4a1345512079f5dcaf4312f8401c7365a4681ea: no such file or directory
        * invalid OCI image tag
```

I think there should be added support for scanning multi-arch images in local OCI layout through one of the options of scanning by digest or scanning by platform. Similar to scanning multi-arch images using docker urls which is already working as in the following commands:

`trivy image --platform linux/arm64 alpine:3.11.0`

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
## Description

Add ability to reference a local OCI image using the digest of it's manifest as in the following example:

```console
$ trivy image --input alpine@sha256:56ae38f2f5c54b98311b8b2463d4861368c451ac17098f4227d84946b42ab96d
```

## Related issues
- Close #3571

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/image/oci.go [hunk 1]
```diff
@@ -10,17 +10,21 @@ import (
...
 
 	// Check if tag is specified in input
 	// e.g. /path/to/oci:0.0.1
-	if strings.Contains(fileName, ":") {
+	if strings.Contains(fileName, "@") {
+		splitFileName := strings.SplitN(fileName, "@", 2)
+		inputFileName = splitFileName[0]
+		inputRef = splitFileName[1]
+	} else if strings.Contains(fileName, ":") {
 		splitFileName := strings.SplitN(fileName, ":", 2)
```
Comment (knqyf263) [inline pkg/fanal/image/oci.go:22]: The code was written before Go 1.18. We can also rewrite it.

```suggestion
		inputFileName, inputRef := strings.Cut(fileName, ":")
```

File: docs/docs/references/configuration/cli/trivy_image.md [hunk 2]
```diff
@@ -15,6 +15,11 @@ trivy image [flags] IMAGE_NAME
   # Scan a container image from a tar archive
   $ trivy image --input ruby-3.1.tar
 
+  # Scan an OCI imag from a local layout referenced by tag or digest
```
Comment (knqyf263) [inline docs/docs/references/configuration/cli/trivy_image.md:18]: This documentation is auto-generated by `mage docs:generate`. Instead, can you please update the following document?
https://github.com/aquasecurity/trivy/blob/7e99d08a131c05adf8d6f70a878c5a943a8f2d7a/docs/docs/target/container_image.md#oci-layout

**Review by knqyf263 — COMMENTED** [review]
It looks good. I left small comments.

File: docs/docs/references/configuration/cli/trivy_image.md (hunk 2 above)
Comment (laurentiuNiculae) [reply docs/docs/references/configuration/cli/trivy_image.md]: Alright, I have moved the examples to the document you linked.

File: pkg/fanal/image/oci.go [hunk 3]
```diff
@@ -10,17 +10,21 @@ import (
...
-	var inputTag, inputFileName string
+	var inputRef, inputFileName string
 
 	// Check if tag is specified in input
 	// e.g. /path/to/oci:0.0.1
-	if strings.Contains(fileName, ":") {
+	if strings.Contains(fileName, "@") {
+		splitFileName := strings.SplitN(fileName, "@", 2)
+		inputFileName = splitFileName[0]
+		inputRef = splitFileName[1]
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/fanal/image/oci.go]: I rewrote the logic using strings.Cut()

File: pkg/fanal/image/oci.go (hunk 1 above)
Comment (laurentiuNiculae) [reply pkg/fanal/image/oci.go]: Yes, I have used the Cut function similarly to check for the tag/digest.

**Review by AnaisUrlichs — APPROVED** [review]

**Comment by knqyf263** [comment]
LGTM. @DmitriyLewen Can you please double-check?

**Review by DmitriyLewen — APPROVED** [review]
I played with changes. It works as expected.
LGTM.

**Comment by knqyf263** [comment]
@DmitriyLewen Thanks!!

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
@laurentiuNiculae Thanks for your contribution!

**Comment by laurentiuNiculae** [comment]
Thank you so much for accepting my change and for the kind and quick response!

**Pull request merged**